Make c10d_functional ops call into _c10d_functional ops

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124992
* __->__ #124979

This PR removes the legacy impls of c10d_functional ops which are now irrelevant. For backward compatibility purpose, c10d_functional ops now call into _c10d_functional ops.

We also changed c10d_functional ops to be CompositeExplicitAutograd, so that when traced, only _c10d_functional ops appear in the graph. After this, we'll be able to remove the Inductor IR for the legacy functional collectives.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Commit:
Make c10d_functional ops call into _c10d_functional ops

[ghstack-poisoned]

Commit:
Update on "Make c10d_functional ops call into _c10d_functional ops"


This PR removes the legacy impls of c10d_functional ops which are now irrelevant. For backward compatibility purpose, c10d_functional ops now call into _c10d_functional ops.

We also changed c10d_functional ops to be CompositeExplicitAutograd, so that when traced, only _c10d_functional ops appear in the graph. After this, we'll be able to remove the Inductor IR for the legacy functional collectives.

[ghstack-poisoned]

Commit:
Update on "Make c10d_functional ops call into _c10d_functional ops"


This PR removes the legacy impls of c10d_functional ops which are now irrelevant. For backward compatibility purpose, c10d_functional ops now call into _c10d_functional ops.

We also changed c10d_functional ops to be CompositeExplicitAutograd, so that when traced, only _c10d_functional ops appear in the graph. After this, we'll be able to remove the Inductor IR for the legacy functional collectives.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -1,289 +1,72 @@
-
-These functions should all be bound to dispatcher ops.  During tracing, the op itself should be
-captured in the graph and the backend should implement the op however it prefers.
+This file contains the op impls for the legacy (c10d_functional) functional collectives.
```
wondering shall we simply delete this file instead 🤔 ?

wanchaol commented on torch/distributed/_functional_collectives_impl.py:
```diff
@@ -1,289 +1,72 @@
-work_version = 0
-
-
-class _WaitRegistration:
```
Nice clean up!!

wanchaol reviewed (approved):
Looking forward to the cleanups! CI failures looks real though

Commit:
Update on "Make c10d_functional ops call into _c10d_functional ops"


This PR removes the legacy impls of c10d_functional ops which are now irrelevant. For backward compatibility purpose, c10d_functional ops now call into _c10d_functional ops.

We also changed c10d_functional ops to be CompositeExplicitAutograd, so that when traced, only _c10d_functional ops appear in the graph. After this, we'll be able to remove the Inductor IR for the legacy functional collectives.

[ghstack-poisoned]

Commit:
Update on "Make c10d_functional ops call into _c10d_functional ops"


This PR removes the legacy impls of c10d_functional ops which are now irrelevant. For backward compatibility purpose, c10d_functional ops now call into _c10d_functional ops.

We also changed c10d_functional ops to be CompositeExplicitAutograd, so that when traced, only _c10d_functional ops appear in the graph. After this, we'll be able to remove the Inductor IR for the legacy functional collectives.

[ghstack-poisoned]

Commit:
Update on "Make c10d_functional ops call into _c10d_functional ops"


This PR removes the legacy impls of c10d_functional ops which are now irrelevant. For backward compatibility purpose, c10d_functional ops now call into _c10d_functional ops.

We also changed c10d_functional ops to be CompositeExplicitAutograd, so that when traced, only _c10d_functional ops appear in the graph. After this, we'll be able to remove the Inductor IR for the legacy functional collectives.

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge

yifuwang commented:
@pytorchbot merge
